crates/moonbuild/src/entry.rs

```diff
@@ -600,15 +602,41 @@ pub fn run_test(
     }
 
     let res = if moonbuild_opt.no_parallelize {
+        let runtime = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()?;
+
         runtime.block_on(async {
             let mut results = vec![];
             for handler in handlers {
+                // Tasks are run sequentially by using the `await` expression directly.
                 results.push(handler.await);
             }
             results
         })
     } else {
-        runtime.block_on(futures::future::join_all(handlers))
+        let thread_count = match std::thread::available_parallelism() {
+            Ok(n) => n.get(),
+            Err(_) => num_cpus::get(),
+        };
+
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .worker_threads(thread_count)
+            .enable_all()
+            .build()?;
+
+        runtime.block_on(async {
+            let mut res_handlers = vec![];
+            for handler in handlers {
+                // Submit tasks to the scheduler
+                res_handlers.push(runtime.spawn(handler));
```

Comment:
do we need another `res_handlers` and `spawn` here? IMO the previous `runtime.block_on(futures::future::join_all(handlers))` is easy to understand

Reply:
When using `runtime.block_on(futures::future::join_all(handlers))`, the threads created by `tokio` are never used. Because the tasks are never spawned/submitted to the scheduler, all of them are run in the current thread. You can debug the code or/and print the thread id in `entry.rs::execute_test` to verify it.